docs/architecture/adr-003-values-propagation.md

```diff
@@ -0,0 +1,416 @@
+and to configure the consensus parameters accordingly—most notably,
+the durations of consensus timeouts.
+
+When value size becomes a relevant factor, it is important to recognize
+that the consensus process comprises two distinct stages:
+
+* **Value Propagation**: Proposed values must be transmitted to all consensus participants, eg system nodes.
+* **Value Decision**: The consensus algorithm selects and decides on one of the successfully propagated values.
+
+The cost of **Value Propagation**, in terms of latency and bandwidth,
+clearly depends on the size of the proposed values. In contrast, the
+**Value Decision** stage is in principle independent of value size and
+incurs roughly a constant cost from the perspective of values.
+
+In Tendermint, **Value Propagation** is performed by default via the
+`PROPOSAL` message. The proposer node broadcasts this message, which
+includes the proposed value `v`.
+
+The **Value Decision** phase involves `PREVOTE` and `PRECOMMIT`
+messages—collectively referred to as *votes*. Each vote includes
+either an identifier `id(v)` of the proposed value `v` or the
+special value `nil`. The function `id(v)` provides a compact
+representation of `v`, typically implemented as a fixed-size hash.
+
+### Problem statement
+
+From the above, we can observe that **Value Propagation** is more challenging,
+as it involves disseminating potentially large amounts of data in
+the network of system nodes. In contrast, the **Value Decision** phase requires only
+the transmission of vote messages, which are relatively small and of
+constant size.
+
+Because value propagation can be a bottleneck, Malachite’s consensus API has to provide
+enough flexibility so that applications can optimize the value propagation stage.
+
+### ADR scope
+
+This ADR documents the design of the Malachite consensus API relevant for
+the **Value Propagation** stage. Specifically, how Malachite enables
+application builders flexibility for bypassing or mitigating the performance
+implications of dealing with value propagation in a large system of nodes.
+
+## Decision
+
+### Building Blocks
+
+There are a number of entities that are involved in the value propagation process:
+
+* **Application:** The software that uses the consensus engine. It is responsible for:
+  * Providing the value to be propagated
+  * Validating received values
+  * Providing data availability for values (both undecided and decided)
+* **Consensus Engine:** The component responsible for managing the interactions between the consensus core and the application, and between the consensus core and the networking layer. For the ones involved in the value propagation process, the consensus engine is responsible for:
+  * Propagating values and/or proposals via the networking layer
+  * Receiving values and/or proposals from the networking layer
+  * Relaying values and/or proposals to the consensus core (pending redesign in #943)
```

Comment:
This issue does not exist.

Reply:
removed as we decided to not wait for the PRs. please check 90a25d6f0f3f9bd7ab3a20c235158291fb91d1c0.